README
Assignment 1.07 allows for parsing of multiple CSV's to fill new structs.

Structures include Pokemon, moves, pokemon_moves, stats, pokemon_stats, experience, and pokemon_types.
This assignment takes an argument given by the command line to determine what CSV to parse.

Compile using make. 
Author: Joseph Jennings (jtj9)

CHANGELOG
commit f370f8c699d4632f39800bff4800b1db9ae22d27
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Apr 6 01:02:01 2023 -0500

    Fixed filePath for pyrite.

commit 01fc085303992c5a597e4bddd9eb490f5ebd02fd
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Apr 6 00:43:58 2023 -0500

    Testing filePath on pyrite.

commit c7f2a3e1afad98957783f589dc8131680298b6f1
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Apr 6 00:41:21 2023 -0500

    Testing functions for pyrite.

commit 4de7a95f6f43b75a6e07539fe71074f98f31c5bd
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Apr 6 00:32:19 2023 -0500

    Added support for the required filepaths.

commit a70a00e00b6f2b849b25223b0134fe172f813523
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Apr 6 00:11:32 2023 -0500

    Added csv parsing for moves, experience, stats,
    and type_names.
    Additionally, added ability to parse input from command line.

commit 4796301d1e514b3fc50e4acb1075c464df1d474c
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Tue Apr 4 22:47:51 2023 -0500

    Added support for parsing all pokemon related CSV's.

commit 341554d34f4d29a249f41dea8a803aa110307427
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Mon Apr 3 21:48:30 2023 -0500

    Merged with Dr. Shaeffer.

commit ae485b873b11b3dd52a9c0464ed5c5226c06b593
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Thu Mar 30 08:07:10 2023 -0500

    Added ability to fly.
    Need to add support for more than 1 char.

commit b8c72204d91d412b27d4beeef66ed2fe8575c891
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 29 23:31:36 2023 -0500

    Updated ReadME

commit 87753b8f2a7472f917f4c5c7a0cba6fe234d923e
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 29 23:30:28 2023 -0500

    Converted C code to C++.

commit 02da322da672d1cc69821822bb3569941b9ac4f1
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 29 21:50:41 2023 -0500

    Merged with Dr. Shaeffer

commit 9751682a6a619fa59894a37845af35931746f3e0
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 22 23:40:31 2023 -0500

    Added trainer battles.

commit 798ba715efb4456d1de595ac26e67b02445f64d4
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 22 22:48:33 2023 -0500

    Updated readme.

commit 5dd83d0a154e652850b3ba7e3f357d744153afb7
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 22 01:20:48 2023 -0500

    Added support for entering and leaving a
    pokemart or center.

commit 3cb3daa511d1fd4598aa8b055771bb19f03077b9
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 22 01:11:43 2023 -0500

    Allowed trainer distance to be updated
    with npc movement.

commit 19ea0afc3ac388e079045681e9a262778e8a2fb1
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Wed Mar 22 00:58:26 2023 -0500

    Added functionality for viewing trainers.
    Also added ability to see how far away they are.

commit 11d3ec72bede2bda3a1a93ab7f754739fd98c6c7
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Tue Mar 21 23:33:56 2023 -0500

    Accidentally commited output file, fixed.

commit 9d2a9c258fa4cb0721ab0e9211120ed0648b9d2f
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Tue Mar 21 23:32:59 2023 -0500

    Added movement for the PC with trainer pathing.
    Additonally added color!

commit 2b5f11cead7078ea5206d13f1097eea9fe617ac4
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Tue Mar 21 23:31:40 2023 -0500

    Added support for PC movement.
    Additonally added color to the screen.

commit 42e9c572fd7111cd9bcacf5344e663f66e5251db
Author: Joseph Jennings <josephjenningsct@gmail.com>
Date:   Tue Mar 21 01:13:12 2023 -0500

    Added lncurses to makefile.
    Added basic window and print to program for ncurses.

commit 9088915d6a21a354b8a4f2a89c9c1f81a603e8f6
Author: JosephTJennings <josephjenningsct@gmail.com>
Date:   Sun Mar 19 21:34:13 2023 -0500

    Merged with Dr. Shaeffer.

commit c335f21bd54b36edebf49edb9e76c0e6fc811884
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Mar 2 03:36:59 2023 -0600

    Added ability to prompt for number of trainers.

commit 4cda4824bc0569b75810a152a7cf333736aa28d5
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Mar 2 03:30:45 2023 -0600

    Added map regeneration...

commit efd297cb49b3f2301a475fb00682820d3abaffb3
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Mar 2 03:06:55 2023 -0600

    Added rival movement.

commit 9cabf55cd9a20c644983acad371e4de4347b43a6
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Mar 2 02:22:45 2023 -0600

    added movement for pacer, wanderer, sentry

commit 47ac26137a223d0a2fa6245bb8597fb809f34965
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Wed Mar 1 01:29:44 2023 -0600

    added struct for npc and added npc spawn functionality.

commit 8fa6c5331401315eb5fdf3ac9c0279ed3386ac45
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Mon Feb 27 15:47:45 2023 -0600

    Merged code with Shaeffers code

commit a4579e7c19aea4c3e953f86e2061f335664d4163
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Feb 23 01:11:18 2023 -0600

    added a function to get the terrain cost at a coordinate

commit fc5612dcb66622b47ce4e693d8f4bc6b7d0fda8e
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Feb 23 00:24:23 2023 -0600

    Partially fixed dijkstras.

commit 00ca3719ec31155016b47f52fe0af8302e93d171
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Thu Feb 23 00:07:19 2023 -0600

    Added diagonal traversal.

commit c38b5297f7265e0a65cc79a1498fba53aeec77ac
Author: josephTJennings <josephjenningsct@gmail.com>
Date:   Wed Feb 22 23:54:31 2023 -0600

    Added cahngelog...

commit 91dd1fc38532075a53229652457bea5cc294419d

io.cpp
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>
#include <stdlib.h>
#include <limits.h>

#include "io.h"
#include "character.h"
#include "poke327.h"

typedef struct io_message {
  /* Will print " --more-- " at end of line when another message follows. *
   * Leave 10 extra spaces for that.                                      */
  char msg[71];
  struct io_message *next;
} io_message_t;

static io_message_t *io_head, *io_tail;

void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void io_reset_terminal(void)
{
  endwin();

  while (io_head) {
    io_tail = io_head;
    io_head = io_head->next;
    free(io_tail);
  }
  io_tail = NULL;
}

void io_queue_message(const char *format, ...)
{
  io_message_t *tmp;
  va_list ap;

  if (!(tmp = (io_message_t *) malloc(sizeof (*tmp)))) {
    perror("malloc");
    exit(1);
  }

  tmp->next = NULL;

  va_start(ap, format);

  vsnprintf(tmp->msg, sizeof (tmp->msg), format, ap);

  va_end(ap);

  if (!io_head) {
    io_head = io_tail = tmp;
  } else {
    io_tail->next = tmp;
    io_tail = tmp;
  }
}

static void io_print_message_queue(uint32_t y, uint32_t x)
{
  while (io_head) {
    io_tail = io_head;
    attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(y, x, "%-80s", io_head->msg);
    attroff(COLOR_PAIR(COLOR_CYAN));
    io_head = io_head->next;
    if (io_head) {
      attron(COLOR_PAIR(COLOR_CYAN));
      mvprintw(y, x + 70, "%10s", " --more-- ");
      attroff(COLOR_PAIR(COLOR_CYAN));
      refresh();
      getch();
    }
    free(io_tail);
  }
  io_tail = NULL;
}

/**************************************************************************
 * Compares trainer distances from the PC according to the rival distance *
 * map.  This gives the approximate distance that the PC must travel to   *
 * get to the trainer (doesn't account for crossing buildings).  This is  *
 * not the distance from the NPC to the PC unless the NPC is a rival.     *
 *                                                                        *
 * Not a bug.                                                             *
 **************************************************************************/
static int compare_trainer_distance(const void *v1, const void *v2)
{
  const character_t *const *c1 = (character_t **)v1;
  const character_t *const *c2 = (character_t **)v2;

  return (world.rival_dist[(*c1)->pos[dim_y]][(*c1)->pos[dim_x]] -
          world.rival_dist[(*c2)->pos[dim_y]][(*c2)->pos[dim_x]]);
}

static character_t *io_nearest_visible_trainer()
{
  character_t **c, *n;
  uint32_t x, y, count;

  c =  (character_t **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = world.cur_map->cmap[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  n = c[0];

  free(c);

  return n;
}

void io_display()
{
  uint32_t y, x;
  character_t *c;

  clear();
  for (y = 0; y < MAP_Y; y++) {
    for (x = 0; x < MAP_X; x++) {
      if (world.cur_map->cmap[y][x]) {
        mvaddch(y + 1, x, world.cur_map->cmap[y][x]->symbol);
      } else {
        switch (world.cur_map->map[y][x]) {
        case ter_boulder:
          attron(COLOR_PAIR(COLOR_MAGENTA));
          mvaddch(y + 1, x, BOULDER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_MAGENTA));
          break;
        case ter_mountain:
          attron(COLOR_PAIR(COLOR_MAGENTA));
          mvaddch(y + 1, x, MOUNTAIN_SYMBOL);
          attroff(COLOR_PAIR(COLOR_MAGENTA));
          break;
        case ter_tree:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, TREE_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_forest:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, FOREST_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_path:
          attron(COLOR_PAIR(COLOR_YELLOW));
          mvaddch(y + 1, x, PATH_SYMBOL);
          attroff(COLOR_PAIR(COLOR_YELLOW));
          break;
        case ter_gate:
          attron(COLOR_PAIR(COLOR_YELLOW));
          mvaddch(y + 1, x, GATE_SYMBOL);
          attroff(COLOR_PAIR(COLOR_YELLOW));
          break;
        case ter_mart:
          attron(COLOR_PAIR(COLOR_BLUE));
          mvaddch(y + 1, x, POKEMART_SYMBOL);
          attroff(COLOR_PAIR(COLOR_BLUE));
          break;
        case ter_center:
          attron(COLOR_PAIR(COLOR_RED));
          mvaddch(y + 1, x, POKEMON_CENTER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_RED));
          break;
        case ter_grass:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, TALL_GRASS_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_clearing:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, SHORT_GRASS_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_water:
          attron(COLOR_PAIR(COLOR_CYAN));
          mvaddch(y + 1, x, WATER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_CYAN));
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          attron(COLOR_PAIR(COLOR_CYAN));
          mvaddch(y + 1, x, ERROR_SYMBOL);
          attroff(COLOR_PAIR(COLOR_CYAN)); 
       }
      }
    }
  }

  mvprintw(23, 1, "PC position is (%2d,%2d) on map %d%cx%d%c.",
           world.pc.pos[dim_x],
           world.pc.pos[dim_y],
           abs(world.cur_idx[dim_x] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_x] - (WORLD_SIZE / 2) >= 0 ? 'E' : 'W',
           abs(world.cur_idx[dim_y] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_y] - (WORLD_SIZE / 2) <= 0 ? 'N' : 'S');
  mvprintw(22, 1, "%d known %s.", world.cur_map->num_trainers,
           world.cur_map->num_trainers > 1 ? "trainers" : "trainer");
  mvprintw(22, 30, "Nearest visible trainer: ");
  if ((c = io_nearest_visible_trainer())) {
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(22, 55, "%c at vector %d%cx%d%c.",
             c->symbol,
             abs(c->pos[dim_y] - world.pc.pos[dim_y]),
             ((c->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              'N' : 'S'),
             abs(c->pos[dim_x] - world.pc.pos[dim_x]),
             ((c->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              'W' : 'E'));
    attroff(COLOR_PAIR(COLOR_RED));
  } else {
    attron(COLOR_PAIR(COLOR_BLUE));
    mvprintw(22, 55, "NONE.");
    attroff(COLOR_PAIR(COLOR_BLUE));
  }

  io_print_message_queue(0, 0);

  refresh();
}

uint32_t io_teleport_pc(pair_t dest)
{
  /* Just for fun. And debugging.  Mostly debugging. */

  do {
    dest[dim_x] = rand_range(1, MAP_X - 2);
    dest[dim_y] = rand_range(1, MAP_Y - 2);
  } while (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]                  ||
           move_cost[char_pc][world.cur_map->map[dest[dim_y]]
                                                [dest[dim_x]]] == INT_MAX ||
           world.rival_dist[dest[dim_y]][dest[dim_x]] < 0);

  return 0;
}

static void io_scroll_trainer_list(char (*s)[40], uint32_t count)
{
  uint32_t offset;
  uint32_t i;

  offset = 0;

  while (1) {
    for (i = 0; i < 13; i++) {
      mvprintw(i + 6, 19, " %-40s ", s[i + offset]);
    }
    switch (getch()) {
    case KEY_UP:
      if (offset) {
        offset--;
      }
      break;
    case KEY_DOWN:
      if (offset < (count - 13)) {
        offset++;
      }
      break;
    case 27:
      return;
    }

  }
}

static void io_list_trainers_display(character_t **c,
                                     uint32_t count)
{
  uint32_t i;
  char (*s)[40]; /* pointer to array of 40 char */

  s = (char (*)[40]) malloc(count * sizeof (*s)); //??

  mvprintw(3, 19, " %-40s ", "");
  /* Borrow the first element of our array for this string: */
  snprintf(s[0], 40, "You know of %d trainers:", count);
  mvprintw(4, 19, " %-40s ", *s);
  mvprintw(5, 19, " %-40s ", "");

  for (i = 0; i < count; i++) {
    snprintf(s[i], 40, "%16s %c: %2d %s by %2d %s",
             char_type_name[c[i]->npc->ctype],
             c[i]->symbol,
             abs(c[i]->pos[dim_y] - world.pc.pos[dim_y]),
             ((c[i]->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              "North" : "South"),
             abs(c[i]->pos[dim_x] - world.pc.pos[dim_x]),
             ((c[i]->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              "West" : "East"));
    if (count <= 13) {
      /* Handle the non-scrolling case right here. *
       * Scrolling in another function.            */
      mvprintw(i + 6, 19, " %-40s ", s[i]);
    }
  }

  if (count <= 13) {
    mvprintw(count + 6, 19, " %-40s ", "");
    mvprintw(count + 7, 19, " %-40s ", "Hit escape to continue.");
    while (getch() != 27 /* escape */)
      ;
  } else {
    mvprintw(19, 19, " %-40s ", "");
    mvprintw(20, 19, " %-40s ",
             "Arrows to scroll, escape to continue.");
    io_scroll_trainer_list(s, count);
  }

  free(s);
}

static void io_list_trainers()
{
  character_t **c;
  uint32_t x, y, count;

  c = (character_t **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = world.cur_map->cmap[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  /* Display it */
  io_list_trainers_display(c, count);
  free(c);

  /* And redraw the map */
  io_display();
}

void io_pokemart()
{
  mvprintw(0, 0, "Welcome to the Pokemart.  Could I interest you in some Pokeballs?");
  refresh();
  getch();
}

void io_pokemon_center()
{
  mvprintw(0, 0, "Welcome to the Pokemon Center.  How can Nurse Joy assist you?");
  refresh();
  getch();
}

void io_battle(character_t *aggressor, character_t *defender)
{
  character_t *npc;

  io_display();
  mvprintw(0, 0, "Aww, how'd you get so strong?  You and your pokemon must share a special bond!");
  refresh();
  getch();
  if (aggressor->pc) {
    npc = defender;
  } else {
    npc = aggressor;
  }

  npc->npc->defeated = 1;
  if (npc->npc->ctype == char_hiker || npc->npc->ctype == char_rival) {
    npc->npc->mtype = move_wander;
  }
}

uint32_t move_pc_dir(uint32_t input, pair_t dest)
{
  dest[dim_y] = world.pc.pos[dim_y];
  dest[dim_x] = world.pc.pos[dim_x];

  switch (input) {
  case 1:
  case 2:
  case 3:
    dest[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    dest[dim_y]--;
    break;
  }
  switch (input) {
  case 1:
  case 4:
  case 7:
    dest[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    dest[dim_x]++;
    break;
  case '>':
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_mart) {
      io_pokemart();
    }
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_center) {
      io_pokemon_center();
    }
    break;
  }

  if (world.cur_map->map[dest[dim_y]][dest[dim_x]] == ter_gate) {
    /* Can't leave the map */  
    return 1;
  }

  if (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]) {
    if (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]->npc &&
        world.cur_map->cmap[dest[dim_y]][dest[dim_x]]->npc->defeated) {
      // Some kind of greeting here would be nice
      return 1;
    } else if (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]->npc) {
      io_battle(&world.pc, world.cur_map->cmap[dest[dim_y]][dest[dim_x]]);
      // Not actually moving, so set dest back to PC position
      dest[dim_x] = world.pc.pos[dim_x];
      dest[dim_y] = world.pc.pos[dim_y];
    }
  }
  
  if (move_cost[char_pc][world.cur_map->map[dest[dim_y]][dest[dim_x]]] ==
      INT_MAX) {
    return 1;
  }

  return 0;
}

void io_handle_input(pair_t dest)
{
  uint32_t turn_not_consumed;
  int key, x, y;

  do {
    switch (key = getch()) {
    case '7':
    case 'y':
    case KEY_HOME:
      turn_not_consumed = move_pc_dir(7, dest);
      break;
    case '8':
    case 'k':
    case KEY_UP:
      turn_not_consumed = move_pc_dir(8, dest);
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      turn_not_consumed = move_pc_dir(9, dest);
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      turn_not_consumed = move_pc_dir(6, dest);
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      turn_not_consumed = move_pc_dir(3, dest);
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      turn_not_consumed = move_pc_dir(2, dest);
      break;
    case '1':
    case 'b':
    case KEY_END:
      turn_not_consumed = move_pc_dir(1, dest);
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      turn_not_consumed = move_pc_dir(4, dest);
      break;
    case '5':
    case ' ':
    case '.':
    case KEY_B2:
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    case '>':
      turn_not_consumed = move_pc_dir('>', dest);
      break;
    case 'Q':
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      world.quit = 1;
      turn_not_consumed = 0;
      break;
      break;
    case 't':
      io_list_trainers();
      turn_not_consumed = 1;
      break;
    case 'p':
      /* Teleport the PC to a random place in the map.              */
      io_teleport_pc(dest);
      turn_not_consumed = 0;
      break;
    case 'f':
      x = getch();
      y = getch();
      if (x >= -(WORLD_SIZE / 2) && x <= WORLD_SIZE / 2 &&
          y >= -(WORLD_SIZE / 2) && y <= WORLD_SIZE / 2) {
        world.cur_idx[dim_x] = x + (WORLD_SIZE / 2);
        world.cur_idx[dim_y] = y + (WORLD_SIZE / 2);
        new_map(1);
      } 
    case 'm':
      
    case 'q':
      /* Demonstrate use of the message queue.  You can use this for *
       * printf()-style debugging (though gdb is probably a better   *
       * option.  Not that it matters, but using this command will   *
       * waste a turn.  Set turn_not_consumed to 1 and you should be *
       * able to figure out why I did it that way.                   */
      io_queue_message("This is the first message.");
      io_queue_message("Since there are multiple messages, "
                       "you will see \"more\" prompts.");
      io_queue_message("You can use any key to advance through messages.");
      io_queue_message("Normal gameplay will not resume until the queue "
                       "is empty.");
      io_queue_message("Long lines will be truncated, not wrapped.");
      io_queue_message("io_queue_message() is variadic and handles "
                       "all printf() conversion specifiers.");
      io_queue_message("Did you see %s?", "what I did there");
      io_queue_message("When the last message is displayed, there will "
                       "be no \"more\" prompt.");
      io_queue_message("Have fun!  And happy printing!");
      io_queue_message("Oh!  And use 'Q' to quit!");

      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    default:
      /* Also not in the spec.  It's not always easy to figure out what *
       * key code corresponds with a given keystroke.  Print out any    *
       * unhandled key here.  Not only does it give a visual error      *
       * indicator, but it also gives an integer value that can be used *
       * for that key in this (or other) switch statements.  Printed in *
       * octal, with the leading zero, because ncurses.h lists codes in *
       * octal, thus allowing us to do reverse lookups.  If a key has a *
       * name defined in the header, you can use the name here, else    *
       * you can directly use the octal value.                          */
      mvprintw(0, 0, "Unbound key: %#o ", key);
      turn_not_consumed = 1;
    }
    refresh();
  } while (turn_not_consumed);
}
